lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -212,7 +245,9 @@ void ThreadPoolExecutor::run(const DAGNode *root,
   // done here instead of inside executeDAGNode() so that a node can be
   // executed while placeholders are being propagated for the next node without
   // the callback for that node deleting the execution state.
-  executionState->incrementInflightNodes((root->children).size());
+  auto numChildren = (root->children).size();
+  executionState->incrementInflightNodes(numChildren);
+  inflightBarrier_.increment(numChildren);
```

Comment:
You want to increment the inflight barrier first before doing anything else, because that is what's stopping the class from being deleted.